### ISSUE TITLE
Possible deadlock bug between Slab rebalancer and Dispatcher thread

### ISSUE THREAD

**Issue by uditagarwal97** [memcached/memcached#738]
Hi @dormando ,

It looks like I've come across a deadlock bug in Memcached. Please find the following bug report. Let me know if you need any more information or in case you want me to open a PR. Thanks.

Regards,
Udit Agarwal

Description of the bug 
---------------------- 

Commit id: 8335dd84ca85651115d518d7e3d963de9d82e22f, Branch: Master

It's a deadlock bug between the dispatcher thread and slab rebalancer thread. Consider the following sequence of events:
   1. Client(s) issues a 'slabs reassign 12 18' command followed by a "shutdown" command to MC.
   2. For processing the 'slab reassign' command, a worker thread will grab the 'slabs_rebalance_lock,' and in do_slabs_reassign() function, it will set the slab_rebalance_signal to 1 and signal the 'slab_rebalance_cond' conditional variable[2]. After the worker thread releases the 'slabs_rebalance_lock,' slab rebalancer thread is expected to stop waiting on the 'slab_rebalance_cond'[3] conditional variable and start moving the slabs, right? However, what if the OS scheduler doesn't immediately schedule the slab rebalancer thread? It will still be waiting on 'slab_rebalance_cond'.
   3. Now, as a consequence of the "shutdown" command, the dispatcher thread will call 'stop_slab_maintenance_thread()' function, in which it will grab the 'slabs_rebalance_lock', set 'do_run_slab_rebalance_thread' to 0, signal 'slab_rebalance_cond' and release the 'slabs_rebalance_lock'[4].
   4. Now if the slab rebalancer thread gets scheduled, it will still try to rebalance the slabs as 'slab_rebalance_signal' is equal to 1 and the condition of the while loop is[1]: "while (slab_rebalance_signal || do_run_slab_rebalance_thread)". Note that at this point the value of do_run_slab_rebalance_thread is 0.
   5. After rebalancing the slabs, 'slab_rebalance_signal' will be equal to 0, and slab rebalancer will again wait on 'slab_rebalance_cond'. While at this point, dispatcher thread is trying to join slab rebalancer using pthread_join[5]. Thus, a deadlock.

The above-mentioned sequence of events will lead to the following interleaving:

Worker thread ---->
     pthread_mutex_lock(&slabs_rebalance_lock)      ../slabs.c: slabs_reassign(int, int)
     ...
     slab_rebalance_signal = 1;                     ../slabs.c: do_slabs_reassign(int, int)
     pthread_cond_signal(&slab_rebalance_cond)
     pthread_mutex_unlock(&slabs_rebalance_lock)    ../slabs.c: slabs_reassign(int, int)

Dispatcher thread ----->
     mutex_lock(&slabs_rebalance_lock);             ../slabs.c: stop_slab_maintenance_thread(void)
     do_run_slab_rebalance_thread = 0;
     pthread_cond_signal(&slab_rebalance_cond);
     pthread_mutex_unlock(&slabs_rebalance_lock);
     pthread_join(rebalance_tid, NULL);    <---------- Dispatcher thread will get blocked here

Slab rebalancer thread ----->
     pthread_cond_wait(&slab_rebalance_cond, &slabs_rebalance_lock);  ../slabs.c: slab_rebalance_thread(void*)
     ...
     pthread_cond_wait(&slab_rebalance_cond, &slabs_rebalance_lock);  <------- Slab rebalncer will get blocked here

Check the possible fixes section to get more clarity of the buggy code.

My System and OS configuration
------------------------------

x86_64, Intel i5, 4 logical cores running Ubuntu 16.

Steps to reproduce this bug 
--------------------------- 

Consider the following exemplary test case (using Cache::Memcached client library):

~~~~

#!/usr/bin/perl
use Test::More tests => 12;
use MemcachedTest;

# Turning off the LRU threads and assoc maintainer thread will significantly improve the probability of triggering the bug
my $server = new_memcached('-m 2 -t 2 -M -o hashpower=16,slab_reassign,no_lru_crawler,no_lru_maintainer,no_hashexpand');
my $sock1 = $server->sock;
my $sock2 = $server->new_sock;

print $sock1 "slabs automove 0\r\n";
my $body = scalar(<$sock1>);
my $expected = "OK\r\n";

if ($body ne $expected) {
    print STDERR "Something terrible has happened: $expected\nBODY:\n$body\nDONETEST\n";
}

my $value_long = "B"x1024x12; # 12KB
my $value_short = "B"x1024x5; # 5KB

my $keycount = 80;
for (1 .. $keycount) {
    print $sock1 "set nfoo$_ 0 0 12288 noreply\r\n$value_long\r\n";
}

$keycount = 140;
for (81 .. $keycount) {
    print $sock1 "set nfoo$_ 0 0 5120 noreply\r\n$value_short\r\n";
}

print $sock1 "slabs reassign 23 19\r\n";
$body = scalar(<$sock1>);

$keycount = 150;
for (141 .. $keycount) {
    print $sock1 "set nfoo$_ 0 0 5120 noreply\r\n$value_short\r\n";
}

print $sock1 "quit\r\n";

$keycount = 350;
for (300 .. $keycount) {
    print $sock2 "set nfoo$_ 0 0 12288 noreply\r\n$value_long\r\n";
}

print $sock2 "shutdown\r\n";

~~~~

The above test case will lead to the deadlock; however, this bug, just like other concurrency bugs, is hard to reproduce; you might need to run the above test case several hundred times. A much faster way is to use gdb along with the above test case. In gdb, you can set the 'scheduler-locking' property to 1.
This will lock the native OS scheduler, and now you can manually enforce the thread interleavings described above. This bug is hard to get so, please feel free to ping me if you need some more details.

Possible fixes 
--------------- 

1. One way is to change the condition of the while loop[1] in slab rebalancer thread from: 'while (slab_rebalance_signal || do_run_slab_rebalance_thread)' to 'while (do_run_slab_rebalance_thread)'. I don't really understand why do we want to let the slab rebalancer run after 'do_run_slab_rebalance_thread' is set to 0.
2. Another easy way is to change the blocking condition of slab rebalancer from 'if (slab_rebalance_signal == 0)' to 'if (slab_rebalance_signal == 0 && do_run_slab_rebalance_thread)'. This will ensure that slab rebalancer thread will wait on the conditional variable only if it is still allowed to run.

References 
----------- 

[1] https://github.com/memcached/memcached/blob/master/slabs.c#L1209
[2] https://github.com/memcached/memcached/blob/master/slabs.c#L1292
[3] https://github.com/memcached/memcached/blob/master/slabs.c#L1232
[4] https://github.com/memcached/memcached/blob/master/slabs.c#L1335
[5] https://github.com/memcached/memcached/blob/master/slabs.c#L1340

Remarks
-------

This behavior was detected by a systematic concurrency tool being developed at Microsoft Research.

**Comment by dormando** [comment]
Thanks!

This one's pretty low priority since it looks super unlikely, and ony affects shutdown. I'll come back to it in the next round of development, or someone else can fix.
